certbot/compat/filesystem.py

```diff
@@ -309,6 +311,105 @@ def is_executable(path):
     return _win_is_executable(path)
 
 
+def is_word_reachable(path):
+    # type: (str) -> bool
+    """
+    Check if everybody/world has any right (read/write/execute) to a file given its path
+    :param str path: path to test
+    :return: True if everybody/world has any right to the file
+    :rtype: bool
+    """
+    if POSIX_MODE:
+        return bool(stat.S_IMODE(os.stat(path).st_mode) & stat.S_IRWXO)
+
+    if not os.path.isfile(path):
+        return False
+
+    security = win32security.GetFileSecurity(path, win32security.DACL_SECURITY_INFORMATION)
+    dacl = security.GetSecurityDescriptorDacl()
+
+    return bool(dacl.GetEffectiveRightsFromAcl({
+        'TrusteeForm': win32security.TRUSTEE_IS_SID,
+        'TrusteeType': win32security.TRUSTEE_IS_USER,
+        'Identifier': win32security.ConvertStringSidToSid('S-1-1-0'),
+    }))
+
+
+def compute_private_key_mode(old_key, base_mode):
+    # type: (str, int) -> int
+    """
+    Calculate the POSIX mode to apply to a private key given the previous private key
+    :param str old_key: path to the previous private key
+    :param int base_mode: the minimum modes to apply to a private key
+    :return: the POSIX mode to apply
+    :rtype: int
+    """
+    if POSIX_MODE:
+        # On Linux, we keep read/write/execute permissions
+        # for group and read permissions for everybody.
+        old_mode = (stat.S_IMODE(os.stat(old_key).st_mode) &
+                    (stat.S_IRGRP | stat.S_IWGRP | stat.S_IXGRP | stat.S_IROTH))
+        return base_mode | old_mode
+
+    # On Windows, the mode returned by os.stat is not reliable,
+    # so we do not keep any permission from the previous private key.
+    return base_mode
+
+
+def get_ownership(path):
+    # type: (str) -> Union[Tuple[int, int], Tuple[str, None]]
+    """
+    Retrieve ownership data about a file given its path
+    :param str path: path to the file
+    :return: on Linux a tuple (UID, GID) of respectively user and group owner of the file,
+             on Windows a tuple (SID, None) where SID corresponds to the user owner of the file.
+    :rtype: tuple of int and int or tuple of str and None
+
+    """
+    if POSIX_MODE:
+        stats = os.stat(path)
+        return stats.st_uid, stats.st_gid
+
+    security = win32security.GetFileSecurity(path, win32security.OWNER_SECURITY_INFORMATION)
+    user = security.GetSecurityDescriptorOwner()
+
+    return win32security.ConvertSidToStringSid(user), None
+
+
+def has_min_permissions(path, min_mode):
```

Comment:
design nit: I feel like `is_word_reachable`, `compute_private_key_mode`, and `has_min_permissions` are all very specific and only work for the current use case.

What do you think instead about a more general function like `get_mode`? On Windows, it would compute the equivalent numeric mode based on the ACEs for the file owner and everyone.

Reply:
It would be the cleanest thing to do, but I do not think it is possible, or at least I think that the result would not be reliable in a security context.

The problem is that the relationship between the POSIX mode and the DACL is not bijective. The DACL is an opiniated approximation of the POSIX mode in the security context.

Take an hyptothetic `get_mode`. There is no group rights on Windows. Should we return the 0 or 7 for the group? Let's assume we compare the octal mode with a reference one, 0o740. Say that `get_mode` returned 0o700 on the first approach, and 0o770 for the second: how do you relate that for Windows, this mode represents the same permissions than the reference?

Let's suppose now that a given file has, for the owner, all flags describing the read permissions, but only some flags of the write permissions. Should the value returned be 4, because not all write permissions are set, or 6 because more than read permissions are set?

The thing is that you cannot manipulate reliably on octal mode representation for Windows. Or you choose some conventions for the examples I gave, but then future developers need to be aware of them, or you limit the risk with... dedicated functions to compare, that are essentially the functions added here. 

So I prefer to not let the caller manipulate the mode, and give meaningful dedicated methods expressing high-level concepts suitable for both systems.

Yes, it adds functions, but we do not add regularly new logic on file permissions in certbot usually. In a year, only one (the private key permissions for everyone+group) has been added.

Reply:
To keep GH up-to-date, Adrien and I talked about this more out of band and he convinced me. My message explaining my thinking here was:

> I was trying to avoid having us to create a bunch of these things whenever we want to inspect the mode in a slightly different way.
>
> I think you convinced me though, by me rereading your comment on GH actually. We have defined a one-to-one mapping from mode to ACL for Certbot's purposes. The inverse is not true and we cannot map an arbitrary ACL to a mode, even if we ignore the group permission problems for a second.
>
> When we encounter something that doesn't fit this model, this deviation cannot be communicated to the caller through the mode so we have either throw away the information or raise an exception. Having multiple special purpose functions that don't try to shove everything into a mode allows more flexibility and lets you more accurately check what you actually care about.